target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-whw03v2.dts

```diff
@@ -0,0 +1,555 @@
+
+    soc {
+        ess_tcsr@1953000 {
+            compatible = "qcom,tcsr";
+            reg = <0x1953000 0x1000>;
+            qcom,ess-interface-select = <TCSR_ESS_PSGMII>;
+        };
+
+
+        tcsr@1949000 {
+            compatible = "qcom,tcsr";
+            reg = <0x1949000 0x100>;
+            qcom,wifi_glb_cfg = <TCSR_WIFI_GLB_CFG>;
+        };
+
+        tcsr@194b000 {
+            compatible = "qcom,tcsr";
+            reg = <0x194b000 0x100>;
+            qcom,usb-hsphy-mode-select = <TCSR_USB_HSPHY_HOST_MODE>;
+        };
+
+        tcsr@1957000 {
+            compatible = "qcom,tcsr";
+            reg = <0x1957000 0x100>;
+            qcom,wifi_noc_memtype_m0_m2 = <TCSR_WIFI_NOC_MEMTYPE_M0_M2>;
+        };
+    };
+
+
+    keys {
+        compatible = "gpio-keys";
+
+        reset {
+            label = "reset";
+            gpios = <&tlmm 18 GPIO_ACTIVE_LOW>;
+            linux,code = <KEY_RESTART>;
+        };
+    };
+};
+
+
+&tlmm {
+    mdio_pins: mdio_pinmux {
+        mux_1 {
+            pins = "gpio6";
+            function = "mdio";  
+            bias-pull-up;
+        };
+
+        mux_2 {
+            pins = "gpio7";
+            function = "mdc";
+            bias-pull-up;
+        };
+    };
+
+    i2c_0_pins: i2c_0_pinmux {
+        mux {
+            function = "blsp_i2c0";
+            pins = "gpio20", "gpio21";
+            bias-disable;
+        };
+    };
+
+    serial_0_pins: serial0_pinmux {
+        mux {
+            pins = "gpio16", "gpio17";
+            function = "blsp_uart0";
+            bias-disable;
+        };
+    };
+
+    serial_1_pins: serial1_pinmux {
+        mux {
+            pins = "gpio8", "gpio9", "gpio10", "gpio11";
+            function = "blsp_uart1";
+            bias-disable;
+        };
+    };
+
+    spi_0_pins: spi_0_pinmux {
+        pinmux {
+            function = "blsp_spi0";
+            pins = "gpio13", "gpio14", "gpio15";
+        };
+
+        pinmux_cs {
+            function = "gpio";
+			pins = "gpio12";
+        };
+
+		pinconf {
+			pins = "gpio13", "gpio14", "gpio15";
+			drive-strength = <12>;
+			bias-disable;
+		};
+
+        pinconf_cs {
+            pins = "gpio12";
+            drive-strength = <2>;
+            bias-disable;
+            output-high;
+        };
+    };
+    
+    spi_1_pins: spi_1_pinmux {
+        cs {
+            pins = "gpio45";
+            function = "gpio";
+            bias-pull-up;
+        };
+
+        host_int {
+            pins = "gpio42";
+            function = "gpio";
+            input;
+        };
+
+        mux {
+            pins = "gpio44", "gpio46", "gpio47";
+            function = "blsp_spi1";
+            bias-disable;
+        };
+
+        reset {
+            pins = "gpio49";
+            function = "gpio";
+            output-high;
+        };
+
+        wake {
+            pins = "gpio31";
+            function = "gpio";
+            output-high;
+        };
+    };
+
+    wifi_0_pins: wifi0_pinmux {
```

Comment:
Sorry, but what is this?
Is it enabling an amplifier or?

Reply:
Oh, its for BT coexistance?

Reply:
Yes, that's the coexist pin that is used [here](https://github.com/vtremblay/openwrt/blob/master/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-whw03v2.dts#L532)**

Reply:
Ok, this is nitpicky but can you then change the node name to something like bt-coexist-pins.

Also, underscores are fine in the labels, but not in node names.

Reply:
This one is a little more tricky than I initially thought. I enabled the debug logs and I saw that the driver does handle the btcoex properly (it flags it as disabled and sets the pin to -1).
There's not a lot of example as I'm the first one to try to configure it, so it might take me a bit longer than expected.

Reply:
Driver is kind of weird here, as its doing:
```
static void ath10k_core_fetch_btcoex_dt(struct ath10k *ar)
{
	struct device_node *node;
	u8 coex_support = 0;
	int ret;

	node = ar->dev->of_node;
	if (!node)
		goto out;

	ret = of_property_read_u8(node, "qcom,coexist-support", &coex_support);
	if (ret) {
		ar->coex_support = true;
		goto out;
	}

	if (coex_support) {
		ar->coex_support = true;
	} else {
		ar->coex_support = false;
		ar->coex_gpio_pin = -1;
		goto out;
	}

	ret = of_property_read_u32(node, "qcom,coexist-gpio-pin",
				   &ar->coex_gpio_pin);
	if (ret)
		ar->coex_gpio_pin = -1;

out:
	ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot coex_support %d coex_gpio_pin %d\n",
		   ar->coex_support, ar->coex_gpio_pin);
}
```

So, by some weird logic if it cannot read the property coexist is enabled, but also if its present it will get enabled and pin read

Reply:
It would be good to add a print and see what is the error code that ret has on `of_property_read_u8`

Reply:
After adding a couple of debug logs in the driver, this what I've discovered:

When the `qcom,coext-support` property is not specified, `of_property_read_u8` returns `EINVAL` (which is expected). The support is turned on by default but the pin is not read.

When we set the property to `<1>`, as specified in the doc, `of_property_read_u8` read `0` instead of `1`. That is because of the data format.
By default <1> is written with 32 bits. The problem is that we are trying to read an `u8`, so it needs to be written with 8 bits.

The solution is to force the right format in the DT:
```
    qcom,coexist-support = /bits/ 8 <0x01>;
    qcom,coexist-gpio-pin = <0x34>;
```
The result can be validated using `hexdump`:
Default 32 bits output:
```
0000000 0000 0100
0000004
```
8 bits output:
```
0000000 0001
0000001
```

Then, the driver is able to read the properties properly:
```
[  109.102048] ath10k_ahb a000000.wifi: boot bt-coex: read qcom,coexist-support ret 0 value 1
[  109.102065] ath10k_ahb a000000.wifi: boot bt-coex: enabled
[  109.102078] ath10k_ahb a000000.wifi: boot bt-coex: read qcom,coexist-gpio-pin ret 0 value 52
[  109.102097] ath10k_ahb a000000.wifi: boot coex_support 1 coex_gpio_pin 52
```

Reply:
I wonder if I should send a patch to fix the doc or to send a patch to change `of_property_read_u8` by `of_property_read_u32` in the driver.

Reply:
In my opinion, the driver should be fixed to read it as u32, reading u8 is rather uncommon and is breaking DT bindings.
If you just update the bindings that still leaves all of the existing DTB-s effectively broken

Reply:
I was hesitating, this feature is probably already used by other projects than openwrt. If we change the driver to read it as u32 instead of u8 it could be breaking.
But I agree that this would be the best way to fix it.

Reply:
That's the thing, bindings kind of trump what the driver is doing now.
Currently, the driver is not following the bindings and so if you have a DTS following them it wont work, that is why driver needs to get fixed instead of bindings, I am not sure that changing the bindings would even be accepted

Reply:
yeah ok, I'll do a patch for the driver and update the PR after, thanks!

Reply:
Quick question; I have [submitted the patch](http://lists.infradead.org/pipermail/ath10k/2022-December/014082.html) to the `ath10k` mailinglist and I will also temporarily add it in `package/kernel/ath10k-ct/patches` directory.

Do I need to submit it to `ath10k-ct` too? Or it will be merged from the `ath10k` driver (if the patch is accepted)?

Reply:
Usually you would wait until it gets accepted upstream for ath10k and then make a PR to backport it to ath10k-ct

Reply:
Will do.
In the meanwhile, the patch is in the `patches` folder of the ath10k-ct package.
Thanks !